config-proxy/src/main/java/com/yahoo/vespa/config/proxy/filedistribution/UrlDownloadRpcServer.java

```diff
@@ -72,60 +68,40 @@ private void download(Request req) {
     private void downloadFile(Request req) {
         String url = req.parameters().get(0).asString();
         File downloadDir = new File(rootDownloadDir, urlToDirName(url));
-        if (alreadyDownloaded(downloadDir)) {
+        Downloader downloader = downloader(url);
+        if (downloader.alreadyDownloaded(downloader, downloadDir)) {
             log.log(Level.INFO, "URL '" + url + "' already downloaded");
-            req.returnValues().add(new StringValue(new File(downloadDir, CONTENTS_FILE_NAME).getAbsolutePath()));
+            req.returnValues().add(new StringValue(new File(downloadDir, downloader.fileName()).getAbsolutePath()));
             req.returnRequest();
             return;
         }
 
         try {
-            URL website = new URL(url);
-            HttpURLConnection connection = (HttpURLConnection) website.openConnection();
-            if (connection.getResponseCode() == 200) {
-                log.log(Level.INFO, "Downloading URL '" + url + "'");
-                downloadFile(req, connection, downloadDir);
-            } else {
-                log.log(Level.SEVERE, "Download of URL '" + url + "' got server response: " + connection.getResponseCode());
-                req.setError(HTTP_ERROR, String.valueOf(connection.getResponseCode()));
-            }
+            Optional<File> file = downloader.downloadFile(url, downloadDir);
+            if (file.isPresent())
+                req.returnValues().add(new StringValue(file.get().getAbsolutePath()));
+            else
+                req.setError(DOES_NOT_EXIST, "URL '" + url + "' not found");
+        } catch (RuntimeException e) {
+            logAndSetRpcError(req, url, e, HTTP_ERROR);
         } catch (Throwable e) {
-            log.log(Level.SEVERE, "Download of URL '" + url + "' failed, got exception: " + e.getMessage());
-            req.setError(INTERNAL_ERROR, "Download of URL '" + url + "' internal error: " + e.getMessage());
+            logAndSetRpcError(req, url, e, INTERNAL_ERROR);
         }
         req.returnRequest();
     }
 
-    private static void downloadFile(Request req, HttpURLConnection connection, File downloadDir) throws IOException {
-        long start = System.currentTimeMillis();
-        String url = connection.getURL().toString();
-        Files.createDirectories(downloadDir.toPath());
-        File contentsPath = new File(downloadDir, CONTENTS_FILE_NAME);
-        try (ReadableByteChannel rbc = Channels.newChannel(connection.getInputStream())) {
-            try (FileOutputStream fos = new FileOutputStream((contentsPath.getAbsolutePath()))) {
-                fos.getChannel().transferFrom(rbc, 0, Long.MAX_VALUE);
+    private static Downloader downloader(String url) {
+        return url.startsWith("s3://") ? new S3Downloader() : new UrlDownloader();
```

Comment:
Consider converting this to `URI`, then this could be a `switch` on `uri.getScheme()`. I guess `UrlDownloader` only supports `http`/`https`? Rest should fail.

In `S3Downloader` the bucket name will be `url.getHost()`.

Reply:
Nice, fixed